storage/filesystem/fat/source/FATFileSystem.cpp

```diff
@@ -330,8 +330,14 @@ int FATFileSystem::unmount()
     }
 
     FRESULT res = f_mount(NULL, _fsid, 0);
-    _ffs[_id] = NULL;
-    _id = -1;
+    if (!res) {
+        int err = _ffs[_id]->deinit();
+        if (err) {
+            return err;
+        }
+        _ffs[_id] = NULL;
+        _id = -1;
```

Comment:
These two lines could probably be executed regardless of whether `f_mount` succeeded.